cardano-ledger-test/src/Test/Cardano/Ledger/ModelChain.hs

```diff
@@ -0,0 +1,698 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE EmptyCase #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE DefaultSignatures #-}
+{-# LANGUAGE DerivingVia #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE FunctionalDependencies #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+module Test.Cardano.Ledger.ModelChain where
+
+import Cardano.Ledger.Coin
+import Cardano.Ledger.Era (Crypto, Era)
+import Cardano.Ledger.Shelley.Constraints
+import Cardano.Slotting.Slot hiding (at)
+import Control.Lens
+import Control.Monad
+import Cardano.Ledger.BaseTypes
+import Cardano.Ledger.Keys
+import Cardano.Ledger.SafeHash (HashAnnotated)
+import Cardano.Ledger.SafeHash (hashAnnotated)
+import Cardano.Slotting.EpochInfo.API
+import Control.Monad.State (MonadState(..))
+import Control.Monad.Trans.Class (lift)
+import Data.Default.Class
+import Data.Foldable
+import Data.Group
+import Data.Kind (Type)
+import Data.Proxy
+import Data.Set (Set)
+import Data.Traversable
+import Data.Word (Word64)
+import Numeric.Natural
+import Shelley.Spec.Ledger.API.Genesis
+import Shelley.Spec.Ledger.API.Mempool
+import Shelley.Spec.Ledger.API.Validation
+import Shelley.Spec.Ledger.Address
+import Shelley.Spec.Ledger.Genesis
+import Shelley.Spec.Ledger.LedgerState (NewEpochState)
+import Shelley.Spec.Ledger.STS.EraMapping ()
+import Test.Shelley.Spec.Ledger.Generator.ScriptClass (mkKeyPairs)
+import Test.Shelley.Spec.Ledger.Utils (mkVRFKeyPair)
+import qualified Cardano.Crypto.DSIGN.Class as DSIGN
+import qualified Cardano.Crypto.Hash.Class as Hash
+import qualified Cardano.Ledger.Core as Core
+import qualified Cardano.Ledger.Crypto as C
+import qualified Cardano.Ledger.Era as Era
+import qualified Cardano.Ledger.Val as Val
+import qualified Control.Monad.Except as Except
+import qualified Control.Monad.State.Class as State
+import qualified Control.Monad.Trans.State as State hiding (state)
+import qualified Data.Map as Map
+import qualified Data.Sequence.Strict as StrictSeq
+import qualified Data.Set as Set
+import qualified GHC.Exts as GHC
+import qualified Shelley.Spec.Ledger.LedgerState as LedgerState
+import qualified Shelley.Spec.Ledger.Tx as Shelley
+import Shelley.Spec.Ledger.EpochBoundary (BlocksMade (..))
+import Shelley.Spec.Ledger.Credential
+  ( Credential(..)
+  , StakeCredential
+  )
+import Shelley.Spec.Ledger.TxBody
+  ( DCert(..)
+  , PoolParams(..)
+  , DelegCert(..)
+  , Delegation(..)
+  , PoolCert(..)
+  )
+import qualified Cardano.Crypto.VRF.Class (VerKeyVRF)
+import qualified Shelley.Spec.Ledger.UTxO as UTxO
+import qualified Shelley.Spec.Ledger.TxBody as Shelley
+import Cardano.Ledger.SafeHash (SafeHash)
+
```

Comment:
There are a lot of open imports here, I always prefer to use either explicit imports or qualified ones. That may just be me, though. You don't necessarily have to do this, I can maybe do it later.